_posts/2025-04-30-#31981.md

```diff
@@ -7,20 +7,55 @@ authors: [Sjors]
+
+Alongside this, there is an ongoing project [multi process Bitcoin](https://github.com/bitcoin/bitcoin/issues/28722) which introduced a new Inter-Process Communication (IPC) scheme, which allows the bitcoind process to communicate with other processes. Noting this, Bitcoin Core contributors suggested creating a mining IPC interface such that an external process, a TP, would connect to it and receive push notifications of new block templates as specified in the spec. This template provider would then forward the block to the [Job Declarator](https://github.com/stratum-mining/sv2-spec/blob/main/06-Job-Declaration-Protocol.md).
+
+The mining interface has been implemented in Bitcoin Core with two classes `BlockTemplate` and `Mining`, which have several methods needed by the Sv2 TP, effectively allowing projects like [Sjors TP](https://github.com/Sjors/bitcoin/pull/49) to utilize it.
+
+This approach has had some latency issues compared to when Bitcoin Core directly acts as the TP.
+
+### Motivation
+Bitcoin Core PR [#31981](https://github.com/bitcoin/bitcoin/pull/31981) adds a new method to the mining interface called `checkBlock` that checks whether a block template is valid or not. This check is equivalent to the one done in `getblocktemplate` RPC proposal mode. Using IPC is more performant, as it avoids serializing up to 4 MB of transaction data as JSON.
+
+- The method that performs this check is [`TestBlockValidity`](https://github.com/bitcoin/bitcoin/blob/9efe5466885862bc1fb830524f7ce23e785fcac0/src/rpc/mining.cpp#L724-L750)
 
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+2. Is the new `CheckBlock` interface part of the TP (yes/no)?
+
+3. Can you list some of the checks performed on the block in `getblocktemplate` proposal mode? Does that include checking for a valid proof of work for the block template?
+
+4. Apart from performance reasons mentioned in the PR title, are there any reasons Sv2 would want to use IPC over RPC?
+
+5. What is the key difference between contextual and context-free checks in the block template validity checks?
+
+6. Is there a change in behavior for `TestBlockValidity` after this PR, or is it a pure refactor? If yes, what's the change?
+
+7. One of the reasons for locking `cs_main` in `TestBlockValidity` is to prevent `CheckBlock` race conditions. Can you describe a scenario when that might happen?
+
+8. In [commit 1d029c23](https://github.com/bitcoin-core-review-club/bitcoin/blob/e6e170cf6c67a56b9c14cece66fdc4fab5f3ec6b/src/validation.cpp#L4662), why do we assert that `fchecked` should return false? Is this a potential issue?
```

Comment:
nit: correct capitalization is important for variables, functions, classes, ...
```suggestion
8. In [commit 1d029c23](https://github.com/bitcoin-core-review-club/bitcoin/blob/e6e170cf6c67a56b9c14cece66fdc4fab5f3ec6b/src/validation.cpp#L4662), why do we assert that `fChecked` should return false? Is this a potential issue?
```

Reply:
Done